src/main/java/com/auth0/TransientCookieStore.java

```diff
@@ -0,0 +1,154 @@
+     * @param nonce the value for the nonce cookie. If null, no cookie will be set.
+     * @param sameSite the value for the SameSite attribute on the cookie
+     * @param useLegacySameSiteCookie whether to set a fallback cookie or not
+     */
+    static void storeNonce(HttpServletResponse response, String nonce, SameSite sameSite, boolean useLegacySameSiteCookie) {
+        store(response, StorageUtils.NONCE_KEY, nonce, sameSite, useLegacySameSiteCookie);
+    }
+
+    /**
+     * Gets the value associated with the state cookie and removes it.
+     * @param request the request object
+     * @param response the response object
+     * @param useLegacySameSiteCookie whether to use a fallback cookie or not
+     * @return the value of the state cookie, if it exists
+     */
+    static String getState(HttpServletRequest request, HttpServletResponse response, boolean useLegacySameSiteCookie) {
+        return getOnce(StorageUtils.STATE_KEY, request, response, useLegacySameSiteCookie);
+    }
+
+    /**
+     * Gets the value associated with the nonce cookie and removes it.
+     * @param request the request object
+     * @param response the response object
+     * @param useLegacySameSiteCookie whether to use a fallback cookie or not
+     * @return the value of the nonce cookie, if it exists
+     */
+    static String getNonce(HttpServletRequest request, HttpServletResponse response, boolean useLegacySameSiteCookie) {
+        return getOnce(StorageUtils.NONCE_KEY, request, response, useLegacySameSiteCookie);
+    }
+
+    private static void store(HttpServletResponse response, String key, String value, SameSite sameSite, boolean useLegacySameSiteCookie) {
+        Validate.notNull(response, "response must not be null");
+        Validate.notNull(key, "key must not be null");
+        Validate.notNull(sameSite, "sameSite must not be null");
+
+        if (value == null) {
+            return;
+        }
+
+        boolean sameSiteNone = SameSite.NONE == sameSite;
+
+        String cookie = String.format("%s=%s; HttpOnly; Max-Age=%d; SameSite=%s", key, value, MAX_AGE_SECONDS, sameSite.getValue());
+        if (sameSiteNone) {
+            cookie = cookie.concat("; Secure");
+        }
+
+        // Servlet Cookie API does not yet support setting the SameSite attribute, so just set cookie on header
+        response.addHeader("Set-Cookie", cookie);
+
+        // set legacy fallback cookie (if configured) for clients that won't accept SameSite=None
+        if (sameSiteNone && useLegacySameSiteCookie) {
+            String legacyCookie = String.format("%s=%s; HttpOnly; Max-Age=%d", "_" + key, value, MAX_AGE_SECONDS);
+            response.addHeader("Set-Cookie", legacyCookie);
+        }
+
+    }
+
+    private static String getOnce(String cookieName, HttpServletRequest request, HttpServletResponse response, boolean useLegacySameSiteCookie) {
+        Cookie[] cookies = request.getCookies();
```

Comment:
**Nit:** the word "cookie" appears here so many times that it's difficult to follow the logic. Maybe better var names here to clarify? This could be `requestCookies` ...

Reply:
Agree, will update for more clarity.

Reply:
Updated variable names for increased clarity and readability.